tari-project/tari
# Remove checkpoints from BlockchainDatabase

## Description

An attempt to remove the MMRCheckpoints from the database. 
Some highlights of main changes to review:
1. Storage of MMRPeakData instead of MMRCheckpoints
1. Calculation of MMR roots logic has changed, but outputted values are the same
1. Validation of Spent Inputs has changed
1. Finding link in chain for reorg `find_orphan_chain_tips` has changed to store and maintain the tips in the database.
1. Removal of MemoryDatabase
1. Change of serialization in keys in LMDB
1. All operations in LMDB require a transaction as opposed to creating a transaction implicitly
1. Removal of `for_each_xxxx` functions
1. Removal of `lmdb_vec`
1. `LocalNet` consensus values have been updated to match `ridcully` instead of `rincewind`
1. Removal of `spent_commitments` field in the `HistoricalBlock`. It has been left as an empty vector for compatibility with the current testnet nodes, but should be removed entirely as soon as it can be.
1. A fix to MutableMMR where the size was wrong if the MMRBackend had existing data in it

Areas which have been marked as unimplemented and need to be completed:
1. Pruned mode sync
1. Search for kernel/TXO


NOTE: There have been some updates to tests in #2412 and this branch includes code from that PR

Thanks to @SWvheerden and @sdbondi who I have taken snippets of code from their branches and supplemented code in this PR

## Motivation and Context


Future work: By removing the memory storage in the `BlockchainDatabase` struct we can also remove the `RwLock` on the `BlockchainDatabase`

## How Has This Been Tested?




## Types of changes

* [ ] Feature refactor (No new feature or functional changes, but performance or technical debt improvements)

## Checklist:

* [ ] I'm merging against the `development` branch.
* [ ] I ran `cargo-fmt --all` before pushing.
* [ ] I ran `cargo test` successfully before submitting my PR.
* [ ] I have squashed my commits into a single commit.
